### PR TITLE
Feature/KITOSUDV-149-search-user-api

### PR DESCRIPTION
**Pull request by a-naskret**
## KITOS Pull request template
The following procedure dictates the steps needed before a Pull request can be merged into master.

- [x] **Implement**: 
      *All requirements are implemented and unit tests are green*
      
- [x] **Merge master into branch / rebase with master**: 
      *Make sure you are testing your changes and how they co-exist with the latest version of master*
      
- [x] **Green on integration**: 
      *All integration tests are green on integration*
      
- [x] **Request review**: 
      *Tag whomever you wish to review your code*
      
- [x] **Review completed**: 
      *Reviewer ticks this box when review comments have been submitted*
      
- [x] **Changes**: 
      *PR owner and reviewer agrees on which changes must be made and the changes are committed.*
      
- [x] **Merge master into branch / rebase with master**: 
      *Make sure you are testing your changes and how they co-exist with the latest version of master*
      
- [x] **Green on integration**: 
      *All integration tests are green on integration*

### CODE REVIEW

File: Presentation.Web/Controllers/API/V1/UserController.cs [hunk 1]
```diff
@@ -214,6 +218,29 @@ public HttpResponseMessage GetUsersWithCrossAccess()
...
         }
 
+        /// <summary>
+        /// Returns the users with matching name or email
+        /// </summary>
+        /// <param name="nameOrEmailQuery">Name or email of the user</param>
+        /// <returns>A list of users</returns>
+        [HttpGet]
+        [Route("api/users/search/{nameOrEmailQuery}")]
+        public HttpResponseMessage SearchUsers(string nameOrEmailQuery, [FromUri] BoundedPaginationQuery paginationQuery = null)
```
Comment (mrjsawdk) [inline Presentation.Web/Controllers/API/V1/UserController.cs:228]: USer query parameters and not route parameters for this. Remember to use url encoding and url decoding.

File: Presentation.Web/Controllers/API/V1/UserController.cs [hunk 2]
```diff
@@ -13,10 +13,14 @@
...
 using Core.DomainServices.Extensions;
 using Core.DomainServices.Generic;
+using Core.DomainServices.Queries;
+using Core.DomainServices.Queries.User;
 using Newtonsoft.Json.Linq;
+using Presentation.Web.Extensions;
 using Presentation.Web.Infrastructure.Attributes;
 using Presentation.Web.Models.API.V1;
 using Presentation.Web.Models.API.V1.Users;
+using Presentation.Web.Models.API.V2.Request.Generic.Queries;
```
Comment (mrjsawdk) [inline Presentation.Web/Controllers/API/V1/UserController.cs:23]: Add a bounded query model to the v1 model scope in stead of using the v2 model (so we can let them have their own life)

File: Presentation.Web/Controllers/API/V1/UserController.cs [hunk 3]
```diff
@@ -224,6 +251,16 @@ private static UserWithOrganizationDTO ToUserWithOrgDTO(UserRoleAssociationDTO d
...
         }
 
+        private static IEnumerable<UserWithEmailDTO> ToUserWithEmailDtos(List<User> users)
+        {
+            return users.Select(ToUserWithEmailDto).ToList();
+        }
+
+        private static UserWithEmailDTO ToUserWithEmailDto(User user)
+        {
+            return new(user.Id, user.Name, user.Email);
```
Comment (mrjsawdk) [inline Presentation.Web/Controllers/API/V1/UserController.cs:261]: check if there isn't already an extension method for this... then the "fullname" issue will not be an issue to you

File: Tests.Integration.Presentation.Web/Tools/HttpApi.cs [hunk 4]
```diff
@@ -425,6 +425,7 @@ public static async Task<(int userId, KitosCredentials credentials, Cookie login
                 using var crypto = new CryptoService();
                 var user = x.AsQueryable().ById(userId);
                 user.Password = crypto.Encrypt(password + user.Salt);
+                user.Name = name;
```
Comment (mrjsawdk) [inline Tests.Integration.Presentation.Web/Tools/HttpApi.cs:428]: Consider taking a last name as well

File: Tests.Integration.Presentation.Web/Users/UserTest.cs [hunk 5]
```diff
@@ -201,13 +201,36 @@ public async Task Delete_User_Returns_Forbidden_When_User_Tries_To_Delete_Himsel
             Assert.Equal(HttpStatusCode.Forbidden, deleteResponse.StatusCode);
         }
 
-        private async Task<(Cookie loginCookie, int userId, OrganizationDTO organization, string email)> CreatePrerequisitesAsync(OrganizationRole role)
+        [Fact]
+        public async Task Search_User_Returns_Users_With_Matching_Name_Or_Email()
+        {
+            var userRole = OrganizationRole.User;
+            var testPhrase = "testPhrase";
```
Comment (mrjsawdk) [inline Tests.Integration.Presentation.Web/Users/UserTest.cs:208]: use `A<string>()`

File: Tests.Integration.Presentation.Web/Users/UserTest.cs [hunk 6]
```diff
@@ -201,13 +201,36 @@ public async Task Delete_User_Returns_Forbidden_When_User_Tries_To_Delete_Himsel
...
+            var name1 = CreateName();
+            var name3 = CreateName();
+
+            var (_, userId1, _, _) = await CreatePrerequisitesAsync(userRole, email1, name1);
+            var (_, userId2, _, _) = await CreatePrerequisitesAsync(userRole, email2, testPhrase);
+            var (_, userId3, _, _) = await CreatePrerequisitesAsync(userRole, email3, name3);
+
+            var users = await UserHelper.SearchUsersAsync(testPhrase);
+
+            Assert.Contains(userId1, users.Select(x => x.Id));
```
Comment (mrjsawdk) [inline Tests.Integration.Presentation.Web/Users/UserTest.cs:222]: add a supplementary error text (last argument) to explain why it expected it.

**Review by mrjsawdk — APPROVED** [review]
Small adjustements and then z2go 🚀

File: Presentation.Web/Controllers/API/V1/UserController.cs (hunk 3 above)
Comment (a-naskret) [reply Presentation.Web/Controllers/API/V1/UserController.cs]: I couldn't find any extension method for this mapping, so I created one in the DTOMappingExtensions

Comment (mrjsawdk) [reply Presentation.Web/Controllers/API/V1/UserController.cs]: Great: Please reuse it here: 
image

**Pull request merged**